File: ui/src/data-model.ts
import { FirebaseApp } from "firebase/app"
import { User } from "firebase/auth"
import { createContext } from "react"

export interface Teacher {
	id: string
	user_id: string
	nickname: string
	email_address: string
	created_at: string
	updated_at: string
}

export interface TeacherData extends Teacher {
	classes?: ClassWithStudents[]
	lesson_plans?: LessonPlanWithQuestions[]
	lessons?: LessonWithResponses[]
}

export interface Class {
	id: string
	teacher_email: string
	name: string
	created_at: string
	updated_at: string
}
export interface ClassWithStudents extends Class {
	students: Student[]
}

export interface Student {
	id: string
	class_id: string
	teacher_email: string
	nickname: string
	notes: string
	created_at: string
	updated_at: string
}

export interface LessonPlan {
	id: string
	teacher_email: string
	title: string
	published: boolean
	created_at: string
	updated_at: string
}
export interface LessonPlanWithQuestions extends LessonPlan {
	questions: LessonQuestion[]
}

export interface LessonQuestion {
	id: string
	lesson_plan_id: string
	teacher_email: string
	body_text: string
	created_at: string
	updated_at: string
	// field_of_study: string
	// specific_topic: string
	categorization_guidance?: string
	media_content_urls?: string[]
	context_material_urls?: string[]
}

export interface Lesson {
	id: string // Not a UUID like the others, rather a shorter string for use as a link
	lesson_name: string
	lesson_plan_id: string
	lesson_plan_name: string
	class_id: string
	class_name: string
	teacher_name: string
	teacher_email: string
	responses_locked: boolean
	created_at: string
	updated_at: string
	deleted?: boolean
	student_names_started?: string[]
	class_data?: ClassWithStudents
	lesson_plan?: LessonPlanWithQuestions
	analysis_by_question_id?: Record<string, LessonQuestionAnalysis>
}

export interface LessonQuestionAnalysis {
	question_id: string
	responses_by_category: Record<string, LessonResponse[]>
}

export interface LessonWithResponses extends Lesson {
	responses?: LessonResponse[]
}

export interface LessonResponse {
	id: string
	teacher_email: string
	question_id: string
	lesson_id: string
	student_id: string
	student_name: string
	response_image_base64?: string
	response_text?: string
	analysis?: LessonResponseAnalysis
	created_at: string
	updated_at: string
}

export interface LessonResponseAnalysis {
	response_summary: string
}

/**
 * UI-only data model
 */
export const AppCtx = createContext<AppContext | null>(null)
export interface AppContext {
	firebaseApp: FirebaseApp
	user: User | null | undefined
	callCloudFunction: <ReturnType = void>(endpoint: string, data?: any, authorization?: string) => Promise<ReturnType | null>
	uploadFile: (file: File, destFolder: string) => Promise<string>
}

File: functions/data_model.py
from dataclasses import dataclass
from typing import Dict, List, Optional

@dataclass
class Teacher:
    id: str
    user_id: str
    nickname: str
    email_address: str
    created_at: str
    updated_at: str

@dataclass
class TeacherData(Teacher):
    classes: List['Class'] = None
    lesson_plans: List['LessonPlan'] = None
    lessons: List['Lesson'] = None

@dataclass
class Class:
    id: str
    teacher_email: str
    name: str
    created_at: str
    updated_at: str
    students: Optional[List['Student']] = None

@dataclass
class Student:
    id: str
    class_id: str
    teacher_email: str
    nickname: str
    notes: str
    created_at: str
    updated_at: str

@dataclass
class LessonPlan:
    id: str
    teacher_email: str
    title: str
    published: bool
    created_at: str
    updated_at: str
    questions: Optional[List['LessonQuestion']] = None


@dataclass
class LessonQuestion:
    id: str
    lesson_plan_id: str
    teacher_email: str
    body_text: str
    # field_of_study: str
    # specific_topic: str
    created_at: str
    updated_at: str
    categorization_guidance: Optional[str] = None
    media_content_urls: Optional[List[str]] = None
    context_material_urls: Optional[List[str]] = None



@dataclass
class Lesson:
    id: str  # Not a UUID like the others, rather a shorter string for use as a link
    lesson_name: str
    lesson_plan_id: str
    lesson_plan_name: str
    class_id: str
    class_name: str
    teacher_name: str
    teacher_email: str
    responses_locked: bool
    created_at: str
    updated_at: str
    deleted: Optional[bool] = None
    student_names_started: Optional[List[str]] = None
    class_data: Optional['Class'] = None
    lesson_plan: Optional['LessonPlan'] = None
    responses: Optional[List['LessonResponse']] = None
    analysis_by_question_id: Optional[Dict[str, 'LessonQuestionAnalysis']] = None


@dataclass
class LessonQuestionAnalysis:
    question_id: str
    responses_by_category: Dict[str, List['LessonResponse']]


@dataclass
class LessonResponse:
    id: str
    teacher_email: str
    question_id: str
    lesson_id: str
    student_id: str
    student_name: str
    created_at: str
    updated_at: str
    response_image_base64: Optional[str] = None
    response_text: Optional[str] = None
    analysis: Optional['LessonResponseAnalysis'] = None


@dataclass
class LessonResponseAnalysis:
    response_summary: str
